### DISCUSSION TITLE
Jump to definition in libraries

### DISCUSSION THREAD

**Discussion by gnfisher** [scalameta/nvim-metals#215] [Q&A] [answered]
Hey folks! I have things working but I can't jump to definition in included libraries any longer. I was able to do this with CoC but its not working with nvim's native lsp. For example, jumping to `OptionT.fromOption` definition. It seems like nvim-metals ought to allow me to do this. I am wondering what I'm missing in terms of setup?

I've used [the examples here](https://github.com/scalameta/nvim-metals/discussions/39) to confirm nvim-metals.

Would appreciate any help, thanks!

**Comment by eugenemedvediev** [comment]
Hi Greg Fisher,

Can you point on `fromOption` and execute the following command `:lua vim.lsp.buf.definition()`?
If it works, then it means that you have broken mappings. 
( in vimscript example I see that Chris has this mapping:
`map("n", "<leader>gD", "<cmd>lua vim.lsp.buf.definition()<CR>")`
but in lua it is gd:
`nnoremap <silent> gd          <cmd>lua vim.lsp.buf.definition()<CR>`

**Reply by eugenemedvediev** [reply]
BTW, I've checked and on my machine with lua mappings I am able to jump to `OptionT.fromOption`

**Reply by gnfisher** [reply]
Thanks for the response! My binding is setup properly and I can jump to definition of everything within the project itself. Just not to anything in the libraries I've included.

**Comment by ckipp01** [comment, marked as answer]
Following up with what @eugenemedvediev mentioned. If you are on top of `fromOption` and you execute a `:lua vim.lsp.buf.definition()`, you can you see what is in the logs at `.metals/metals.log` or use `:MetalsToggleLogs`? Do you see anything funky in the logs right when you try to jump?

**Reply by gnfisher** [reply]
Nothing is logged when I try to make the jump.

**Reply by ckipp01** [reply]
Is this happening in all of your projects, or only this one? For example if you try it in an example project jumping to a library does it work? If normal go-to completions are working as expected, then it's odd that external libraries wouldn't unless the sources are missing somehow, but that seems unlikely.

**Reply by gnfisher** [reply]
Thanks, @ckipp01. I'll play around in a different repo and see if the behavior is different and report back.

**Reply by gnfisher** [reply]
Yes, it works in other projects! @ckipp01 I'm wondering what could have been changed when this stopped working some months ago! Any ideas where I might start looking?

**Reply by ckipp01** [reply]
haha, it depends on deep you want to dig. The steps I typically follow in a situation like this is more or less the following:

1. Try <x> and realize it's not working for some reason
2. Take a look at the logs in `.metals/metals.log` to see if there is something funky going on. (most of the time this will show you)
3. If the above doesn't work I try to ensure that the request is actually going to the server and looking at the response. You can do this by creating an `lsp.trace.json` file as outlined [here](https://scalameta.org/metals/docs/contributors/getting-started/#json-rpc-trace). This will show you the full LSP communication between Metals and your client. You'll see, or should see, something like this:

```
[Trace - 10:54:32 PM] Received request 'textDocument/definition - (11)'
Params: {
  "textDocument": {
    "uri": "file:///Users/ckipp/Documents/scala-workspace/sanity/src/main/scala/Main.scala"
  },
  "position": {
    "line": 5,
    "character": 25
  }
}


[Trace - 10:54:33 PM] Sending response 'textDocument/definition - (11)'. Processing request took 181ms
Result: [
  {
    "uri": "file:///Users/ckipp/Documents/scala-workspace/sanity/.metals/readonly/dependencies/src.zip/java/lang/String.java",
    "range": {
      "start": {
        "line": 2831,
        "character": 18
      },
      "end": {
        "line": 2831,
        "character": 29
      }
    }
  }
]
```

If there is no completion request sent into the server, then it points towards something wrong in your mappings, however since it works in a minimal project, then I don't think that's the case.  However I am curious if Metals is just returning `[]` in this case or what's going on.

You could also see if something funky is going on with Bloop by switching to sbt BSP assuming you're using sbt. You can do this by using the `:MetalsSwitchBsp` command and choose sbt.

**Reply by gnfisher** [reply]
This is the trace..

```
[Trace - 05:10:10 PM] Received request 'textDocument/definition - (8)'
Params: {
  "textDocument": {
    "uri": "file:///home/greg/Code/CLIENT/CLIENT-quoting-server/src/main/scala/CLIENT/quoting/messaging/Server.scala"
  },
  "position": {
    "line": 13,
    "character": 22
  }
}


[Trace - 05:10:10 PM] Sending response 'textDocument/definition - (8)'. Processing request took 37ms
Result: []
```

Here's a trace for a jump to something defined in the project itself:
```
[Trace - 05:11:10 PM] Received notification 'metals/didFocusTextDocument'
Params: "file:///home/greg/Code/ADS/ads-quoting-server/src/main/scala/ads/quoting/upload/BulkUploader.scala"


[Trace - 05:11:11 PM] Received request 'textDocument/definition - (9)'
Params: {
  "textDocument": {
    "uri": "file:///home/greg/Code/CLIENT/CLIENT-quoting-server/src/main/scala/CLIENT/quoting/upload/BulkUploader.scala"
  },
  "position": {
    "line": 27,
    "character": 12
  }
}


[Trace - 05:11:11 PM] Sending response 'textDocument/definition - (9)'. Processing request took 51ms
Result: [
  {
    "uri": "file:///home/greg/Code/CLIENT/CLIENT-quoting-server/src/main/scala/CLIENT/quoting/messaging/Server.scala",
    "range": {
      "start": {
        "line": 7,
        "character": 6
      },
      "end": {
        "line": 7,
        "character": 12
      }
    }
  }
]
```

The result that comes back is an empty array.

**Reply by gnfisher** [reply]
I am trying to switch to SBT but I just see "connect to sbt 23s" and it never connects, just keeps ticking up. @ckipp01 

I will dive into this a bit later tonight and see what I can find. Thanks for your help!

**Reply by gnfisher** [reply]
So! I switched back from SBT to Bloop... it reimported the build... and its working!

Maybe something funky left in the cached bloop files?

Thanks for your help @ckipp01

**Reply by ckipp01** [reply]
Glad it's fixed!

> Maybe something funky left in the cached bloop files?

I'm not sure. Typically if that is the case a `:MetalsCompileClean` should take care of that since it should clean your bloop stuff and then re-compile. Plus, if that was the case I'd expect an error about missing semanticdb or something for that specific file you were in. So I'm really unsure what happened here.
